### PR TITLE
LOG-2169: Fix Kafka sasl-plaintext/sasl-ssl

### PR DESCRIPTION
**Pull request by vimalk78**
### Description
Fixes Kafka Sasl for vector.

Tested scenarios
 - sasl
 - sasl + ssl
 - ssl
 - mutual ssl
 - without tls

/cc @alanconway @jcantrill 
/assign @alanconway 

/cherry-pick 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-2169, https://issues.redhat.com/browse/LOG-2167

### CODE REVIEW

File: internal/generator/vector/output/kafka/kafka.go [hunk 1]
```diff
@@ -128,68 +128,69 @@ timestamp_format = "rfc3339"
...
-	if o.Secret != nil {
+	hasTls := false
+	u, _ := url.Parse(o.URL)
+	if urlhelper.IsTLSScheme(u.Scheme) {
 		if security.HasPassphrase(secret) {
 			pp := Passphrase{
 				PassphrasePath: security.SecretPath(o.Secret.Name, constants.Passphrase),
 			}
 			conf = append(conf, pp)
+			hasTls = true
```
Comment (alanconway) [inline internal/generator/vector/output/kafka/kafka.go:144]:
```suggestion
			hasTLS = true
```

File: internal/generator/vector/output/kafka/sasl.go [hunk 2]
```diff
@@ -1,34 +1,29 @@
...
-func (t SaslConf) Template() string {
-	return `
-{{define "kafkaSasl" -}}
+func (t Sasl) Template() string {
+	return `{{define "vectorKafkaSasl"}}
 # {{.Desc}}
 [sinks.{{.ComponentID}}.sasl]
-{{- end}}`
+enabled = true
+username = "{{.Username}}"
```
Comment (alanconway) [inline internal/generator/vector/output/kafka/sasl.go:25]: What happens if username/password are missing/empty? Is `username = ""` the same as no username?

**Review by alanconway — CHANGES_REQUESTED** [review]
One concern - username, password and mechanism are all optional, it looks like this may generate bad config when they are missing - e.g. in vector configuration I don't think `username=""` is equivalent to having no username. Otherwise looks good.

File: internal/generator/vector/output/kafka/sasl.go (hunk 2 above)
Comment (vimalk78) [reply internal/generator/vector/output/kafka/sasl.go:25]: this configuration is mainly for `SASL/PLAIN` , which needs username/password
We could add a verification step to check username/password is always configured in secret, if `SASL/PLAIN` is configured

File: internal/generator/vector/output/kafka/kafka.go [hunk 3]
```diff
@@ -58,8 +58,8 @@ func Conf(o logging.OutputSpec, inputs []string, secret *corev1.Secret, op Optio
 			Output(o, inputs, secret, op),
 			Encoding(o, op),
 		},
-		TLSConf(o, secret),
-		SASLConf(o, secret),
+		TlsConf(o, secret),
+		SaslConf(o, secret),
```
[the remark this answers is not in the document]
Comment (vimalk78) [reply internal/generator/vector/output/kafka/kafka.go]: updated

File: internal/generator/vector/output/kafka/kafka.go (hunk 1 above)
Comment (vimalk78) [reply internal/generator/vector/output/kafka/kafka.go]: updated

**Pull request merged**